File: concurrent_perfect_number_finder.py
import multiprocessing
import time
import sys


def main():
    if len(sys.argv) < 2:
        print "Must enter a number"
        return
    n = 0    
    try:
        n = int(sys.argv[1])    
    except ValueError:
        print "Invalid input"
        return    
    start_time = time.time()
    find_perfect_number(n)
    end_time = time.time()
    print "Running find_perfect_number() in {} seconds".format(end_time-start_time)
    
    start_time = time.time()
    find_perfect_number_in_parallel(n)
    end_time = time.time()
    print "Running find_perfect_number_in_parallel() in {} seconds".format(end_time-start_time)
       

def find_perfect_number(limit):
    print "Perfect numbers from 1 to {}:".format(limit)
    for x in xrange(1, limit+1):
        if is_perfect_number(x):
            print x
            
            
def find_perfect_number_in_parallel(limit): 
    print "Perfect numbers from 1 to {}:".format(limit)           
    pool_size = multiprocessing.cpu_count() * 2
    process_pool = multiprocessing.Pool(processes=pool_size)
    check_perfect_number_results = process_pool.map(is_perfect_number, range(1, limit+1))
    for i, result in enumerate(check_perfect_number_results):
        if result:
            print i+1
    
    
def is_perfect_number(number):
    sum = 0
    for x in xrange(1, number):
        if number % x == 0:
            sum += x
        if sum > number:
            return False
    return sum == number 
   

if __name__=='__main__':
    main()


   
                               

File: ConcurrentPerfectNumberFinder.java
import java.lang.IllegalArgumentException;
import java.util.concurrent.*;


public class ConcurrentPerfectNumberFinder {
	public static void main(String[] args) {
		if (args.length < 1) {
			throw new IllegalArgumentException("Must enter a number");
		}
		
		int n= 0;
		try {
			n = Integer.parseInt(args[0]);
		} catch (NumberFormatException e) {
			System.out.println("Please enter a valid number");
		}
		
		long startTime = System.currentTimeMillis();
		findPerfectNumber(n);
		long endTime = System.currentTimeMillis();
		long runningTime = endTime - startTime;
		System.out.println("Running findPerfectNumber() in " + runningTime + " ms");
		
		
		startTime = System.currentTimeMillis();
		findPerfectNumberInParallel(n);
		endTime = System.currentTimeMillis();
		runningTime = endTime - startTime;
		System.out.println("Running findPerfectNumberInParallel() in " + runningTime + " ms");
	}
	
	
	private static void findPerfectNumber(int n) {
		System.out.println("Perfect numbers from 1 to " + n + ":");
		for (int i=2; i<=n;i++) {
			if (isPerfectNumber(i)) {
				System.out.println(i);
			}
		}
	}
	
	private static void findPerfectNumberInParallel(int n) {
		System.out.println("Perfect numbers from 1 to " + n + ":");
        ExecutorService executor =  Executors.newFixedThreadPool(3);
		
		for (int i=2; i<=n;i++) {
			Callable<Boolean> perfectNumberCheckerRunner = new PerfectNumberChecker(i);
			Future<Boolean> checkResultFuture = executor.submit(perfectNumberCheckerRunner);
			try {
				if (checkResultFuture.get()) {
					System.out.println(i);
				}
			} catch (ExecutionException e) {
			    System.out.println("ExecutionException()");
			} catch (InterruptedException e) {
			    System.out.println("InterruptedException()");				
			} catch (CancellationException e) {
			    System.out.println("CancellationException()");				
			}
		}		
		
		try {
		    executor.shutdown();
		    executor.awaitTermination(5, TimeUnit.SECONDS);
		}
		catch (InterruptedException e) {
		    System.err.println("Tasks interrupted");
		}
		finally {
		    if (!executor.isTerminated()) {
		        System.err.println("Cancel non-finished tasks");
		    }
		    executor.shutdownNow();
		}
		
	}
	
	
	private static boolean isPerfectNumber(int number) {
		int sum = 0;
		for (int i = 1; i < number; i++) {
			if (number % i == 0) {
				sum += i;
			}
			if (sum > number) {
				return false;
			}
		}	
		
		if (sum == number) {
			return true;
		} 
		
		return false;
		
	}
	
}

class PerfectNumberChecker implements Callable<Boolean> {
	private final int number;
	
	public PerfectNumberChecker(int number) {
		this.number = number;
	}
	
	@Override
	public Boolean call() throws Exception {
		int sum = 0;
		for (int i = 1; i < number; i++) {
			if (number % i == 0) {
				sum += i;
			}
			if (sum > number) {
				return false;
			}
		}	
	
		if (sum == number) {
			return true;
		} 
	
		return false;
	}
}